scripts/test.py

```diff
@@ -0,0 +1,212 @@
+]
+
+def fail(msg):
+    print('ERROR: %s' % msg)
+    exit(1)
+
+def test_pattern(pattern, string):
+    return (pattern and (pattern == '*' or pattern == string))
+
+def parse_filter(pattern_src):
+    if not(pattern_src):
+        return (None, None)
+    elif pattern_src == '*':
+        return ('*', '*')
+    else:
+        parts = pattern_src.split(':')
+        if len(parts) != 2:
+            fail('invalid filter syntax')
+        [profile, test] = parts
+        return (profile, test)
+
+def filter_test_permutations(whitelist_filters, blacklist_filters):
+    whitelist_patterns = [parse_filter(filter) for filter in whitelist_filters]
+    blacklist_patterns = [parse_filter(filter) for filter in blacklist_filters]
+
+    def keep(profile, test):
+        whitelisted = all(
+            test_pattern(profile_pat, profile) and test_pattern(test_pat, test)
+            for (profile_pat, test_pat) in whitelist_patterns
+        )
+        blacklisted = any(
+            test_pattern(profile_pat, profile) and test_pattern(test_pat, test)
+            for (profile_pat, test_pat) in blacklist_patterns
+        )
+        return whitelisted and not(blacklisted)
+
+    # too complex to represent as a list expression in python
+    result = {}
+    for (profile,tests) in test_permutations.items():
+        for test in tests:
+            if keep(profile, test):
+                result[profile] = result[profile] if profile in result else []
+                result[profile].append(test)
+    return result
+
+def run(args):
+    # wraps a "wet" action, printing if dry run is set, executing otherwise
+    def wet(msg, f):
+        if args.dry_run:
+            print(msg)
+        else:
+            f()
+
+    def run_cmd(cmd, on_failure):
+        def do():
+            if subprocess.call(['bash', '-c', cmd], stdout=sys.stdout, stderr=sys.stderr) != 0:
+                on_failure()
+        wet('$ ' + cmd, do)
+
+    coda_exe_path = 'src/app/cli/src/coda.exe'
+    coda_build_path = './_build/default'
+    coda_exe = os.path.join(coda_build_path, coda_exe_path)
+
+    jq_filter = '.level=="Warn" or .level=="Error" or .level=="Faulty_peer" or .level=="Fatal"'
+
+    test_permutations = filter_test_permutations([args.test_pattern], [args.blacklist_pattern])
+    if len(test_permutations) == 0:
+        # TODO: support direct test dispatching
+        if args.test_pattern:
+            fail('no tests were selected -- whitelist pattern did not match any known tests')
+        else:
+            fail('no tests were selected -- blacklist is too restrictive')
+
+    print('Preparing to run the following tests:')
+    for (profile, tests) in test_permutations.items():
+        print('- %s:' % profile)
+        for test in tests:
+            print('  - %s' % test)
+
+    timestamp = int(time.time())
+    print('======================================')
+    print('=============', timestamp, '=============')
```

Comment:
this is printing as a tuple.

Reply:
```
======================================
============= 1552440690 =============
======================================
```
looks like that for me

Reply:
oh, it's py2 vs py3.

Reply:
I'll change it to do string concatenation since apparently that's different across versions.